### PR TITLE
LOG-5806: Refactor output templates to use a template remap

### PR DESCRIPTION
**Pull request by Clee2691**
### Description
This PR refactors outputs that support vector's template syntax to use a remap allowing for fallback values in case fields are missing. Functional tests are added for supported outputs testing user defined tenants/indices.

It also changes the API for index for Splunk and Elasticsearch. Index is not required as a field for Splunk as opposed to Elasticsearch.

/cc @cahartma @vparfonov 
/assign @jcantrill 

### Links
- JIRA: https://issues.redhat.com/browse/LOG-5806

### CODE REVIEW

File: api/observability/v1/output_types.go [hunk 1]
```diff
@@ -411,8 +401,20 @@ type Elasticsearch struct {
...
+	// The Index can be a combination of static and dynamic values consisting of field paths followed by `||` followed by another field path or a static value.
+	// A dynamic value is encased in single curly brackets `{}` and MUST end with a static fallback value separated with `||`.
+	// Static values can only contain alphanumeric characters along with dashes, underscores, dots and forward slashes.
+	// Example:
+	// 1. foo-{.bar||"none"}
+	// 2. {.foo||.bar||"missing"}
+	// 3. foo.{.bar.baz||.qux.quux.corge||.grault||"nil"}-waldo.fred{.plugh||"none"}
+	//
+	// +kubebuilder:validation:Pattern:=`^(([a-zA-Z0-9-_.\/])*(\{(\.[a-zA-Z0-9_]+|\."[^"]+")+((\|\|)(\.[a-zA-Z0-9_]+|\.?"[^"]+")+)*\|\|"[^"]*"\})*)*$`
+	// +kubebuilder:validation:Required
```
Comment (jcantrill) [inline api/observability/v1/output_types.go:416]: I believe this is redundant give we have defined a pattern that does not allow spaces

File: api/observability/v1/output_types.go [hunk 2]
```diff
@@ -450,9 +452,16 @@ type GoogleCloudLogging struct {
...
+	// The LogID can be a combination of static and dynamic values consisting of field paths followed by `||` followed by another field path or a static value.
+	// A dynamic value is encased in single curly brackets `{}` and MUST end with a static fallback value separated with `||`.
+	// Static values can only contain alphanumeric characters along with dashes, underscores, dots and forward slashes.
+	// Example:
+	// 1. foo-{.bar||"none"}
+	// 2. {.foo||.bar||"missing"}
+	// 3. foo.{.bar.baz||.qux.quux.corge||.grault||"nil"}-waldo.fred{.plugh||"none"}
+	//
+	// +kubebuilder:validation:Pattern:=`^(([a-zA-Z0-9-_.\/])*(\{(\.[a-zA-Z0-9_]+|\."[^"]+")+((\|\|)(\.[a-zA-Z0-9_]+|\.?"[^"]+")+)*\|\|"[^"]*"\})*)*$`
 	// +kubebuilder:validation:Required
```
Comment (jcantrill) [inline api/observability/v1/output_types.go:464]: same here.  please remove given we are cleaning up the other

File: api/observability/v1/output_types.go [hunk 3]
```diff
@@ -583,10 +592,18 @@ type Kafka struct {
...
+	// Static values can only contain alphanumeric characters along with dashes, underscores, dots and forward slashes.
+	// Example:
+	// 1. foo-{.bar||"none"}
+	// 2. {.foo||.bar||"missing"}
+	// 3. foo.{.bar.baz||.qux.quux.corge||.grault||"nil"}-waldo.fred{.plugh||"none"}
+	//
+	// +kubebuilder:validation:Pattern:=`^(([a-zA-Z0-9-_.\/])*(\{(\.[a-zA-Z0-9_]+|\."[^"]+")+((\|\|)(\.[a-zA-Z0-9_]+|\.?"[^"]+")+)*\|\|"[^"]*"\})*)*$`
+	// +kubebuilder:validation:Optional
+	// +nullable
+	Topic string `json:"topic,omitempty"`
```
Comment (jcantrill) [inline api/observability/v1/output_types.go:606]: Does nullable and optional really allow us to disregard the regex?

Do we:
* have a test that confirms the api server will except this either missing or empty?
* We should talk about the "initializers" and in another PR set this to our "default" if the previous bullet is true.

File: api/observability/v1/output_types.go [hunk 4]
```diff
@@ -697,12 +714,20 @@ type Loki struct {
...
+	// Static values can only contain alphanumeric characters along with dashes, underscores, dots and forward slashes.
+	// Example:
+	// 1. foo-{.bar||"none"}
+	// 2. {.foo||.bar||"missing"}
+	// 3. foo.{.bar.baz||.qux.quux.corge||.grault||"nil"}-waldo.fred{.plugh||"none"}
+	//
+	// +kubebuilder:validation:Pattern:=`^(([a-zA-Z0-9-_.\/])*(\{(\.[a-zA-Z0-9_]+|\."[^"]+")+((\|\|)(\.[a-zA-Z0-9_]+|\.?"[^"]+")+)*\|\|"[^"]*"\})*)*$`
+	// +kubebuilder:validation:Optional
+	// +nullable
+	TenantKey string `json:"tenantKey,omitempty"`
```
Comment (jcantrill) [inline api/observability/v1/output_types.go:730]: same here

File: api/observability/v1/output_types.go [hunk 5]
```diff
@@ -739,8 +764,21 @@ type Splunk struct {
...
+	// Static values can only contain alphanumeric characters along with dashes, underscores, dots and forward slashes.
+	// Example:
+	// 1. foo-{.bar||"none"}
+	// 2. {.foo||.bar||"missing"}
+	// 3. foo.{.bar.baz||.qux.quux.corge||.grault||"nil"}-waldo.fred{.plugh||"none"}
+	//
+	// +kubebuilder:validation:Pattern:=`^(([a-zA-Z0-9-_.\/])*(\{(\.[a-zA-Z0-9_]+|\."[^"]+")+((\|\|)(\.[a-zA-Z0-9_]+|\.?"[^"]+")+)*\|\|"[^"]*"\})*)*$`
+	// +kubebuilder:validation:Optional
+	// +nullable
+	Index string `json:"index,omitempty"`
```
Comment (jcantrill) [inline api/observability/v1/output_types.go:781]: same

File: internal/generator/vector/output/elasticsearch/elasticsearch.go [hunk 6]
```diff
@@ -53,6 +54,7 @@ func New(id string, o obs.OutputSpec, inputs []string, secrets helpers.Secrets,
 			Debug(id, helpers.MakeInputs(inputs...)),
 		}
 	}
+	componentID := helpers.MakeID(id, "elasticsearch_index")
```
Comment (jcantrill) [inline internal/generator/vector/output/elasticsearch/elasticsearch.go:57]: nit.  you could probably just call this index.  I defer to you if you wish to alter it, but not a blocker for me

**Review by jcantrill — COMMENTED** [review]
/approve

File: api/observability/v1/output_types.go (hunk 3 above)
Comment (Clee2691) [reply api/observability/v1/output_types.go:606]: I'm not sure about `nullable`, but  I've seen this designation in the API for when things are marked as optional so I threw it in there. I know that `optional` will disregard the regex when `omitempty` is not defined. But `omitempty` by itself will disregard the regex.

File: api/observability/v1/output_types.go (hunk 4 above)
Comment (Clee2691) [reply api/observability/v1/output_types.go:730]: Same reply as above

File: api/observability/v1/output_types.go (hunk 5 above)
Comment (Clee2691) [reply api/observability/v1/output_types.go:781]: Same reply as above

**Review by jcantrill — APPROVED** [review]
/lgtm

**Pull request merged**